=== modified file 'conf/Makefile.common'
--- conf/Makefile.common	2012-03-10 15:55:48 +0000
+++ conf/Makefile.common	2012-04-02 12:42:14 +0000
@@ -11,10 +11,10 @@
   CFLAGS_PLATFORM += -mrtd -mregparm=3
 endif
 if COND_i386_efi
-  LDFLAGS_PLATFORM = -melf_i386
+  LDFLAGS_PLATFORM = -Wl,-melf_i386
 endif
 if COND_x86_64_efi
-  LDFLAGS_PLATFORM = -melf_x86_64
+  LDFLAGS_PLATFORM = -Wl,-melf_x86_64
 endif
 if COND_i386_qemu
   CFLAGS_PLATFORM += -mrtd -mregparm=3
@@ -35,7 +35,7 @@
 endif
 if COND_sparc64_ieee1275
   CFLAGS_PLATFORM += -mno-app-regs
-  LDFLAGS_PLATFORM = -melf64_sparc -mno-relax
+  LDFLAGS_PLATFORM = -Wl,-melf64_sparc -mno-relax
 endif
 
 # Other options
