=== modified file 'grub-core/kern/x86_64/efi/callwrap.S'
--- grub-core/kern/x86_64/efi/callwrap.S	2009-07-02 00:21:07 +0000
+++ grub-core/kern/x86_64/efi/callwrap.S	2011-03-31 10:44:40 +0000
@@ -37,80 +37,80 @@
         .text
 
 FUNCTION(efi_wrap_0)
-	subq $40, %rsp
+	subq $48, %rsp
 	call *%rdi
-	addq $40, %rsp
+	addq $48, %rsp
 	ret
 
 FUNCTION(efi_wrap_1)
-	subq $40, %rsp
+	subq $48, %rsp
 	mov  %rsi, %rcx
 	call *%rdi
-	addq $40, %rsp
+	addq $48, %rsp
 	ret
 
 FUNCTION(efi_wrap_2)
-	subq $40, %rsp
+	subq $48, %rsp
 	mov  %rsi, %rcx
 	call *%rdi
-	addq $40, %rsp
+	addq $48, %rsp
 	ret
 
 FUNCTION(efi_wrap_3)
-	subq $40, %rsp
+	subq $48, %rsp
 	mov  %rcx, %r8
 	mov  %rsi, %rcx
 	call *%rdi
-	addq $40, %rsp
+	addq $48, %rsp
 	ret
 
 FUNCTION(efi_wrap_4)
-	subq $40, %rsp
+	subq $48, %rsp
 	mov %r8, %r9
 	mov %rcx, %r8
 	mov %rsi, %rcx
 	call *%rdi
-	addq $40, %rsp
+	addq $48, %rsp
 	ret
 
 FUNCTION(efi_wrap_5)
-	subq $40, %rsp
+	subq $48, %rsp
 	mov %r9, 32(%rsp)
 	mov %r8, %r9
 	mov %rcx, %r8
 	mov %rsi, %rcx
 	call *%rdi
-	addq $40, %rsp
+	addq $48, %rsp
 	ret
 
 FUNCTION(efi_wrap_6)
-	subq $56, %rsp
-	mov 56+8(%rsp), %rax
+	subq $64, %rsp
+	mov 64+8(%rsp), %rax
 	mov %rax, 40(%rsp)
 	mov %r9, 32(%rsp)
 	mov %r8, %r9
 	mov %rcx, %r8
 	mov %rsi, %rcx
 	call *%rdi
-	addq $56, %rsp
+	addq $64, %rsp
 	ret
 
 FUNCTION(efi_wrap_10)
-	subq $88, %rsp
-	mov 88+40(%rsp), %rax
+	subq $96, %rsp
+	mov 96+40(%rsp), %rax
 	mov %rax, 72(%rsp)
-	mov 88+32(%rsp), %rax
+	mov 96+32(%rsp), %rax
 	mov %rax, 64(%rsp)
-	mov 88+24(%rsp), %rax
+	mov 96+24(%rsp), %rax
 	mov %rax, 56(%rsp)
-	mov 88+16(%rsp), %rax
+	mov 96+16(%rsp), %rax
 	mov %rax, 48(%rsp)
-	mov 88+8(%rsp), %rax
+	mov 96+8(%rsp), %rax
 	mov %rax, 40(%rsp)
 	mov %r9, 32(%rsp)
 	mov %r8, %r9
 	mov %rcx, %r8
 	mov %rsi, %rcx
 	call *%rdi
-	addq $88, %rsp
+	addq $96, %rsp
 	ret
