diff --git a/util/grub-install.in b/util/grub-install.in
index c0d7d8b..1d3158d 100644
--- a/util/grub-install.in
+++ b/util/grub-install.in
@@ -78,8 +78,6 @@ else
     disk_module=native
 fi
 
-. "${datadir}/@PACKAGE@/grub-mkconfig_lib"
-
 # Usage: usage
 # Print the usage.
 usage () {
@@ -126,6 +124,8 @@ echo
 gettext "Report bugs to <bug-grub@gnu.org>."; echo
 }
 
+. "${datadir}/@PACKAGE@/grub-mkconfig_lib"
+
 argument () {
   opt="$1"
   shift
diff --git a/util/grub-kbdcomp.in b/util/grub-kbdcomp.in
index 2be541b..48acc94 100644
--- a/util/grub-kbdcomp.in
+++ b/util/grub-kbdcomp.in
@@ -12,8 +12,6 @@ grub_mklayout="${bindir}/`echo grub-mklayout | sed ${transform}`"
 
 ckbcomp_options=""
 
-. "${datadir}/@PACKAGE@/grub-mkconfig_lib"
-
 self=`basename $0`
 
 usage () {
@@ -29,6 +27,8 @@ usage () {
     gettext "Report bugs to <bug-grub@gnu.org>."; echo
 }
 
+. "${datadir}/@PACKAGE@/grub-mkconfig_lib"
+
 argument () {
   opt=$1
   shift
diff --git a/util/grub-mkconfig.in b/util/grub-mkconfig.in
index 859c2e8..db61522 100644
--- a/util/grub-mkconfig.in
+++ b/util/grub-mkconfig.in
@@ -42,8 +42,6 @@ grub_script_check="${bindir}/`echo grub-script-check | sed "${transform}"`"
 
 GRUB_PREFIX=`echo '/@bootdirname@/@grubdirname@' | sed "s,//*,/,g"`
 
-. "${datadir}/@PACKAGE@/grub-mkconfig_lib"
-
 # Usage: usage
 # Print the usage.
 usage () {
@@ -57,6 +55,8 @@ usage () {
     gettext "Report bugs to <bug-grub@gnu.org>."; echo
 }
 
+. "${datadir}/@PACKAGE@/grub-mkconfig_lib"
+
 argument () {
   opt=$1
   shift
diff --git a/util/grub-mknetdir.in b/util/grub-mknetdir.in
index c96167d..7a8c483 100644
--- a/util/grub-mknetdir.in
+++ b/util/grub-mknetdir.in
@@ -52,8 +52,6 @@ efi32_dir="${libdir}/$(echo ${PACKAGE_TARNAME} | sed ${transform})/i386-efi"
 efi64_dir="${libdir}/$(echo ${PACKAGE_TARNAME} | sed ${transform})/x86_64-efi"
 itanium_dir="${libdir}/$(echo ${PACKAGE_TARNAME} | sed ${transform})/ia64-efi"
 
-. "${datadir}/@PACKAGE@/grub-mkconfig_lib"
-
 # Usage: usage
 # Print the usage.
 usage () {
@@ -72,6 +70,8 @@ usage () {
     gettext "Report bugs to <bug-grub@gnu.org>."; echo
 }
 
+. "${datadir}/@PACKAGE@/grub-mkconfig_lib"
+
 argument () {
   opt=$1
   shift
diff --git a/util/grub-mkrescue.in b/util/grub-mkrescue.in
index e080064..2f5ebca 100644
--- a/util/grub-mkrescue.in
+++ b/util/grub-mkrescue.in
@@ -44,9 +44,7 @@ rom_directory=
 override_dir=
 grub_mkimage="${bindir}/`echo grub-mkimage | sed ${transform}`"
 
-xorriso=xorriso
-
-. "${datadir}/@PACKAGE@/grub-mkconfig_lib"
+xorriso="xorriso"
 
 # Usage: usage
 # Print the usage.
@@ -70,6 +68,8 @@ usage () {
     gettext "Mail xorriso support requests to <bug-xorriso@gnu.org>."; echo
 }
 
+. "${datadir}/@PACKAGE@/grub-mkconfig_lib"
+
 argument () {
   opt=$1
   shift
diff --git a/util/grub-mkstandalone.in b/util/grub-mkstandalone.in
index 52a045d..6a775e0 100644
--- a/util/grub-mkstandalone.in
+++ b/util/grub-mkstandalone.in
@@ -40,8 +40,6 @@ format=
 grub_mkimage="${bindir}/`echo grub-mkimage | sed ${transform}`"
 source=
 
-. "${datadir}/@PACKAGE@/grub-mkconfig_lib"
-
 # Usage: usage
 # Print the usage.
 usage () {
@@ -67,6 +65,8 @@ usage () {
     gettext "Report bugs to <bug-grub@gnu.org>."; echo
 }
 
+. "${datadir}/@PACKAGE@/grub-mkconfig_lib"
+
 argument () {
   opt=$1
   shift
diff --git a/util/grub-reboot.in b/util/grub-reboot.in
index 400f497..74df3f3 100644
--- a/util/grub-reboot.in
+++ b/util/grub-reboot.in
@@ -34,8 +34,6 @@ rootdir=
 bootdir=
 grubdir=`echo "/@bootdirname@/@grubdirname@" | sed 's,//*,/,g'`
 
-. "${datadir}/@PACKAGE@/grub-mkconfig_lib"
-
 # Usage: usage
 # Print the usage.
 usage () {
@@ -52,6 +50,8 @@ usage () {
     gettext "Report bugs to <bug-grub@gnu.org>."; echo
 }
 
+. "${datadir}/@PACKAGE@/grub-mkconfig_lib"
+
 argument () {
   opt=$1
   shift
diff --git a/util/grub-set-default.in b/util/grub-set-default.in
index 0010027..9bcc4f3 100644
--- a/util/grub-set-default.in
+++ b/util/grub-set-default.in
@@ -34,8 +34,6 @@ rootdir=
 bootdir=
 grubdir=`echo "/@bootdirname@/@grubdirname@" | sed 's,//*,/,g'`
 
-. "${datadir}/@PACKAGE@/grub-mkconfig_lib"
-
 # Usage: usage
 # Print the usage.
 usage () {
@@ -52,6 +50,8 @@ usage () {
     gettext "Report bugs to <bug-grub@gnu.org>."; echo
 }
 
+. "${datadir}/@PACKAGE@/grub-mkconfig_lib"
+
 argument () {
   opt=$1
   shift
