=== modified file 'conf/any-emu.rmk'
--- conf/any-emu.rmk	2010-02-03 00:24:07 +0000
+++ conf/any-emu.rmk	2010-02-12 16:46:14 +0000
@@ -10,6 +10,7 @@
 	commands/handler.c commands/ls.c commands/test.c 		\
 	commands/search_wrap.c commands/search_file.c			\
 	commands/search_label.c commands/search_uuid.c			\
+	commands/search_part_uuid.c						\
 	commands/blocklist.c commands/hexdump.c				\
 	lib/hexdump.c commands/halt.c commands/reboot.c			\
 	lib/envblk.c commands/loadenv.c					\

