diff --git a/DuetPkg/PostBuild.bat b/DuetPkg/PostBuild.bat
index 6d71aeb..a995d17 100644
--- a/DuetPkg/PostBuild.bat
+++ b/DuetPkg/PostBuild.bat
@@ -25,7 +25,7 @@
 @if "%1"=="X64" set PROCESSOR=X64
 @if %PROCESSOR%=="" goto WrongArch
 
-@set BUILD_DIR=%WORKSPACE%\Build\DuetPkg%PROCESSOR%\DEBUG_MYTOOLS
+@set BUILD_DIR=%WORKSPACE%\Build\DuetPkg%PROCESSOR%\DEBUG_VS2008x86
 
 
 @echo Compressing DUETEFIMainFv.FV ...
